lib/common/zstd_deps.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2016-2020, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under both the BSD-style license (found in the
+ * LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ * in the COPYING file in the root directory of this source tree).
+ * You may select, at your option, one of the above-listed licenses.
+ */
+
+/* Need:
+ * memcpy()
+ * memset()
+ * memmove()
+ * BYTE
+ * S16
+ * U16
+ * U32
+ * U64
+ * size_t
+ * ptrdiff_t
+ * INT_MAX
+ * ...
+ */
+#ifndef ZSTD_DEPS_COMMON
+#define ZSTD_DEPS_COMMON
+
+#include <limits.h>
+#include <stddef.h>
+#include <string.h>
```

Comment:
This seems to include `<string.h>` everywhere, 
notably in numerous places where only `<stddef.h>` was previously required.

This feels slightly overkill.

`string.h` is notably useful for `memcpy()` and `memset()`, and occasionally for additional functions.
I suspect these places are not that numerous.
Typically, I expect this situation to happen in some `*.c` units, and (generally) not in `*.h` headers.

Reply:
`"mem.h"` currently includes `<string.h>`, so that means it is basically everywhere. `zstd_internal.h` needs memcpy.

Reply:
`mem.h` is indeed included in many `*.c` units,
but I believe we paid attention to not include it unduly into most `*.h` headers.

Reply:
You have already paid attention to not include "everything" when including `zstd_deps.h`
thanks to the `ZSTD_DEPS_NEED_XXX` pattern.

I guess it could be extended to `<string.h>`.

In which case, `mem.h` could include `zstd_deps.h` with `ZSTD_DEPS_NEED_STRING` (for example)
and that would make it equivalent to what we have today,
with `<string.h>` indirectly included when `mem.h` is included.

Note that I do not extend this request to `<limits.h>`.
`<limits.h>` only contains a few constants, so it's harmless.
`<mem.h>` in contrast includes many functions in an expanded "ready-to-inline" format, so it's an entirely different payload.

Reply:
All but a few headers in `common/` like `debug.h` get it. Most headers end up including one of `zstd_internal.h`, `zstd_compress_internal.h`, or `zstd_decompress_internal.h`. All of which indirectly include `<string.h>`.

I don't think the gains of not including `<string.h>` in a small number of `common/` headers pay for the cost of having to `#define ZSTD_DEPS_NEED_STRING` before nearly every include of `zstd_deps.h`.

The important thing is that we don't depend on `zstd_deps.h` in our public headers `zstd.h` and `zstd_errors.h`, which we don't.

Reply:
I'm not against a cleanup for headers that only need `size_t`. In those cases we don't need `zstd_deps.h`, instead we can include `<stddef.h>` like `zstd.h` does.

I'm also not against a cleanup of `zstd_*_internal.h` to split it up into smaller pieces that are only types/declarations that don't need `memcpy()`, and thus don't need `zstd_deps.h`.

But that isn't the current state of the repo, so I think that can happen in a separate PR.

Reply:
> I don't think the gains of not including `<string.h>` in a small number of `common/` headers pay for the cost of having to `#define ZSTD_DEPS_NEED_STRING` before nearly every include of `zstd_deps.h`.

I agree this wouldn't be a good trade-off.
But I was under the perhaps misleading assumption that this would only be necessary in `mem.h`, since most other files would include it through this channel.

Reply:
> But I was under the perhaps misleading assumption that this would only be necessary in mem.h, since most other files would include it through this channel.

That would work. Though I'm a fan of including what you use in the file directly, and not depending on transitive includes.